src/scheduler/pbs_bitmap.c

```diff
@@ -0,0 +1,346 @@
+	long prev_longs;
+	
+	if(num_bits <= 0)
+		return NULL;
+	
+	if(pbm == NULL) {
+		bm = malloc(sizeof(pbs_bitmap));
+		if(bm == NULL)
+			return NULL;
+		bm->bits = NULL;
+		bm->num_longs = 0;
+	}
+	else 
+		bm = pbm;
+	
+	/* If we have enough unused bits available, we don't need to allocate */
+	if (bm->num_longs * BYTES_TO_BITS(sizeof(unsigned long)) >= num_bits) {
+		bm->num_bits = num_bits;
+		return bm;
+	}
+		
+		
+	prev_longs = bm->num_longs;
+	
+	bm->num_bits = num_bits;
+	bm->num_longs = num_bits / BYTES_TO_BITS(sizeof(unsigned long));
+	if (num_bits % BYTES_TO_BITS(sizeof(unsigned long)) > 0)
+		bm->num_longs++;
+	tmp_bits = calloc(bm->num_longs, sizeof(unsigned long));
+	if (tmp_bits == NULL) {
+		if(pbm == NULL) /* we allocated the memory */
+			pbs_bitmap_free(bm);
+		return NULL;
+	}
+	
+	if(bm->bits != NULL) {
+		int i;
+		for(i = 0; i < prev_longs; i++)
+			tmp_bits[i] = bm->bits[i];
+		
+		free(bm->bits);
+	}
+	bm->bits = tmp_bits;
+		
+	return bm;
+}
+
+/* pbs_bitmap destructor */
+void
+pbs_bitmap_free(pbs_bitmap *bm)
+{
+	if(bm == NULL)
+		return;
+	free(bm->bits);
+	free(bm);
+}
+
+/**
+ * @brief turn a bit on for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - which bit to turn on
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_on(pbs_bitmap *pbm, long bit)
+{
+	long n;
+	unsigned long b;
+	unsigned long f;
+	
+	if(pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	n = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	f = bit % BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << f;
+	
+	pbm->bits[n] |= b;
+	return 1;
+}
+
+/**
+ * @brief turn a bit off for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - the bit to turn off
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_off(pbs_bitmap *pbm, long bit)
+{
+	long n;
+	unsigned long b;
+	
+	if (pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	n = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
+	
+	pbm->bits[n] &= ~b;
+	return 1;
+}
+
+/**
+ * @brief get the value of a bit
+ * @param pbm - the bitmap
+ * @param bit - which bit to get the value of
+ * @return int
+ * @retval 1 if the bit is on
+ * @retval 0 if the bit is off
+ */
+int
+pbs_bitmap_get_bit(pbs_bitmap *pbm, unsigned long bit)
+{
+	long n;
+	unsigned long b;
+	
+	if (pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits)
+		return 0;
+	
+	n = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
+	
+	return (pbm->bits[n] & b) ? 1 : 0;
+}
+
+/**
+ * @brief starting at a bit, get the next on bit
+ * @param pbm - the bitmap
+ * @param start_bit - which bit to start from
+ * @return int
+ * @retval number of next on bit
+ * @retval -1 if there isn't a next on bit
+ */
+int
+pbs_bitmap_get_next_on_bit(pbs_bitmap *pbm, long start_bit)
+{
+	long long_ind;
```

Comment:
I like these identifier names much better than 'n' and 'b', can you please rename the same variables in other functions  to these?

Reply:
OK, I renamed n, but left b.  I named n long_ind.  It is an index into the long array.  I didn't want to rename b to bit_ind since it isn't an index.  I couldn't call it bit since there is a parameter named bit.  I just left it 'b'.  I also renamed local_bit_ind in find_next_on_bit() to bit.